src/pmacFilterControl.cpp

```diff
@@ -625,16 +625,20 @@ void PMACFilterController::_transition_state(ControlState state)
             _get_time(&this->last_process_ts_);
         }
         bool waiting = state == ControlState::WAITING || state == ControlState::SINGLESHOT_WAITING;
+
+        // If moving to an Error or IDLE state, or moving to a waiting state from an Error or IDLE state and
+        // not at max attenuation, set max attenuation
+        {
+        if (state < 1 || (this->state_ != ControlState::SINGLESHOT_COMPLETE && waiting && this->current_attenuation_ != MAX_ATTENUATION))
+            this->_set_attenuation(MAX_ATTENUATION);
+        }
+
         // If currently in a scan and current state is waiting, or the scan is now complete, reset singleshot start flag
         if (
             (this->singleshot_start_ && this->state_ == ControlState::WAITING) || state == ControlState::SINGLESHOT_COMPLETE)
         {
             this->singleshot_start_ = false;
         }
-        if (state < 1 || (waiting && this->state_ >= 0))
-        {
-            this->_set_attenuation(MAX_ATTENUATION);
-        }
     }
```

Comment:
You could maybe just change the existing case in place to make the diff clearer as they have ended up being pretty similar. Just add a check that it is not currently in singleshot complete. It could also be tidied up a bit as adding conditions is making it hard to read:

```suggestion
        // Set max attenuation if
        bool set_max_attenuation = false;
        //   - Moving to an idle or error state
        set_max_attenuation |= state < 1;
        //   - or first entering a waiting state, i.e. not from a previously completed singleshot run
        set_max_attenuation |= (waiting && this->state_ != ControlState::SINGLESHOT_COMPLETE)
        
        if (set_max_attenuation)
        {
            this->_set_attenuation(MAX_ATTENUATION);
        }
    }
```

You could also add a `set_max_attenuation &= this->current_attenuation_ != MAX_ATTENUATION;`, but I don't know if that is really necessary. Is it possible to hit this transition when already at max attenuation?